### PR TITLE
Improve account lock for Email OTP for failed attempts.

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Improve account lock for Email OTP for failed attempts.

### CODE REVIEW

**Review by inthirakumaaran — APPROVED** [review]

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 1]
```diff
@@ -432,7 +450,10 @@ private boolean isBackUpCodeValid(String[] savedOTPs, String userToken) {
      */
     private boolean isBackupCodeEnabled(AuthenticationContext context) {
 
-        return EmailOTPUtils.getConfiguration(context, EmailOTPAuthenticatorConstants.BACKUP_CODE).equals("true");
+        boolean isLocalUser = isLocalUser(context);
+        boolean isBackupCodeEnabled = "true".equals(EmailOTPUtils.getConfiguration(context,
+                EmailOTPAuthenticatorConstants.BACKUP_CODE));
+        return isLocalUser && isBackupCodeEnabled;
```
Comment (tharindu-b-hewage) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]:
```suggestion
        return isLocalUser(context) && "true".equals(EmailOTPUtils.getConfiguration(context,
                EmailOTPAuthenticatorConstants.BACKUP_CODE));
```

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 2]
```diff
@@ -2100,4 +2121,215 @@ private UserRealm getUserRealm(String username) throws AuthenticationFailedExcep
...
+        }
+    }
+
+    /**
+     * Check whether the user being authenticated via a local authenticator or not.
+     *
+     * @param context Authentication context.
+     * @return Whether the user being authenticated via a local authenticator.
+     */
+    public static boolean isLocalUser(AuthenticationContext context) {
```
Comment (tharindu-b-hewage) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: can't we make this a private method?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 3]
```diff
@@ -2100,4 +2121,215 @@ private UserRealm getUserRealm(String username) throws AuthenticationFailedExcep
...
+            } else if (EmailOTPAuthenticatorConstants.PROPERTY_ACCOUNT_LOCK_TIME.equals(connectorConfig.getName()) &&
+                    NumberUtils.isNumber(connectorConfig.getValue())) {
+                unlockTimePropertyValue = Integer.parseInt(connectorConfig.getValue());
+            } else if (EmailOTPAuthenticatorConstants.PROPERTY_LOGIN_FAIL_TIMEOUT_RATIO
+                    .equals(connectorConfig.getName()) && NumberUtils.isNumber(connectorConfig.getValue())) {
+                double value = Double.parseDouble(connectorConfig.getValue());
+                if (value > 0) {
+                    unlockTimeRatio = value;
+                }
+            }
```
Comment (tharindu-b-hewage) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Shall we use `switch` statement with config name tom  improve readability

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 4]
```diff
@@ -2100,4 +2121,215 @@ private UserRealm getUserRealm(String username) throws AuthenticationFailedExcep
...
+        }
+    }
+
+    /**
+     * Execute account lock flow for OTP verification failures.
+     *
+     * @param context Authentication context.
+     * @throws AuthenticationFailedException Exception on authentication failure.
+     */
+    private void handleOtpVerificationFail(AuthenticationContext context) throws AuthenticationFailedException {
```
Comment (tharindu-b-hewage) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2210]: If possible, separate logic in this method to several private methods....

Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2210]: fixed in https://github.com/wso2-extensions/identity-outbound-auth-email-otp/pull/79/commits/1fee67f5a09229baa2729e473e5d0ee839c07795

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 3 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Fixed in 1fee67f

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 2 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Fixed in 1fee67f

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 1 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Fixed in 1fee67f

**Review by tharindu-b-hewage — APPROVED** [review]

**Pull request merged**